xla/client/lib/math.cc

```diff
@@ -1175,10 +1175,130 @@ XlaOp Acos(XlaOp x) {
+          break;
+        case F32:
+          safe_min_ = std::sqrt(std::numeric_limits<float>::min()) * 4;
+          safe_max_ = std::sqrt(std::numeric_limits<float>::max()) / 8;
+          break;
+        default:
+          safe_min_ = safe_max_ = 0.0;
+          assert(false);  // unreachable
+      }
+      auto signed_x = Real(z);
+      auto signed_y = Imag(z);
+      auto x = Abs(signed_x);
+      auto y = Abs(signed_y);
+
+      auto safe_min = ScalarLike(x, safe_min_);
+      auto safe_max = ScalarLike(x, safe_max_);
+      auto safe_max_m6 = ScalarLike(x, safe_max_ * 1e-6);
+      auto safe_max_p12 = ScalarLike(x, safe_max_ * 1e12);
+      auto safe_max_p2 = ScalarLike(x, safe_max_ * 1e2);
+      auto safe_max_opt = Select(Lt(x, safe_max_p12), safe_max_m6, safe_max_p2);
+      auto y_ge_max_opt = Ge(y, safe_max_opt);
+
+      auto zero = ScalarLike(x, 0);
+      auto one = ScalarLike(x, 1);
+      auto none = ScalarLike(x, -1);
+      auto half = ScalarLike(x, 0.5);
+      auto log2 = ScalarLike(x, std::log(2.0));
+      auto xp1 = Add(x, one);
+      auto xm1 = Add(x, none);
+      auto yy = Mul(y, y);
+      auto half_yy = Mul(half, yy);
+      auto r = Hypot(xp1, y);
+      auto s = Hypot(xm1, y);
+      auto a = Mul(half, Add(r, s));
+      auto ap1 = Add(a, one);
+      auto half_apx = Mul(half, Add(a, x));
+      auto rpxp1 = Add(r, xp1);
+      auto spxm1 = Add(s, xm1);
+      auto smxm1 = Sub(s, xm1);
+      auto y_is_finite = Not(IsPosInf(y));
+      auto max_xy = Select(Gt(x, y), x, y);
+
+      auto y1 = Select(
+          Gt(max_xy, safe_max), y,
+          Select(
+              Le(x, one), Sqrt(Mul(half_apx, Add(Div(yy, rpxp1), smxm1))),
+              Mul(y, Sqrt(Add(Div(half_apx, rpxp1), Div(half_apx, spxm1))))));
+      auto real = Atan2(signed_x, y1);
+
+      auto am1 = Select(
+          And(Lt(y, safe_min), Lt(x, one)), Mul(none, Div(Mul(xp1, xm1), ap1)),
+          Select(Ge(x, one), Add(Div(half_yy, rpxp1), Mul(half, spxm1)),
+                 Select(Le(a, ScalarLike(a, 1.5)),
+                        Add(Div(half_yy, rpxp1), Div(half_yy, smxm1)),
+                        Add(a, none))));
+      auto sq = Sqrt(Mul(am1, ap1));
+      auto xoy = Select(And(y_ge_max_opt, y_is_finite), Div(x, y), zero);
+      auto log1p_xoy = Log1p(Mul(xoy, xoy));
+      auto mx = Select(y_ge_max_opt, y, x);
+      auto imag = Select(Ge(mx, Select(y_ge_max_opt, safe_max_opt, safe_max)),
+                         Add(log2, Add(Log(mx), Mul(half, log1p_xoy))),
+                         Select(And(Lt(y, safe_min), Lt(x, one)), Div(y, sq),
+                                Log1p(Add(am1, sq))));
+      auto signed_imag = Select(Lt(signed_y, zero), Mul(none, imag), imag);
+
+      return Complex(real, signed_imag);
+    }
+    // TODO: handle overflow in x*x
+    return ScalarLike(z, 2.0) *
+           Atan2(x, ScalarLike(z, 1.0) + Sqrt(ScalarLike(z, 1.0) - z * z));
```

Comment:
```suggestion
           Atan2(z, ScalarLike(z, 1.0) + Sqrt(ScalarLike(z, 1.0) - z * z));
```

Reply:
Thanks for catching the typo! This actually indicates that arcus sine on real inputs is not tested in XLA.

Reply:
> This actually indicates that arcus sine on real inputs is not tested in XLA.

Actually it does not because `z` and `x` are essentially the same objects.

Reply:
Fixed.